### PR TITLE
[WIP] String Concat Function Implementation

### PR DESCRIPTION
**Pull request by xiazcy**
I've opened a draft PR in the TinkerPop main repo with this new branch. It was unfortunately tracked with the JIRA ticket I put into the title, so I'm opening this PR here so we can comment without flooding people's email.

### CODE REVIEW

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConcatGlobalStep.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+import org.apache.tinkerpop.gremlin.process.traversal.Traversal;
+import org.apache.tinkerpop.gremlin.process.traversal.Traverser;
+import org.apache.tinkerpop.gremlin.process.traversal.step.TraversalParent;
+import org.apache.tinkerpop.gremlin.process.traversal.traverser.TraverserRequirement;
+
+import java.util.Collections;
+import java.util.Set;
+
+public class ConcatGlobalStep<S, E> extends ScalarMapStep<S, String> implements TraversalParent {
```
Comment (xiazcy) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConcatGlobalStep.java:29]: My question here is if we need to bound the parameters of the step? That would let the runtime throw the casting error for arguments and we can catch and throw illegal argument from that, instead of allowing anything then checking type and throwing illegal argument.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConcatStep.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        return this.isAllNull? null : sb.toString();
+    }
+
+    private String processTraversal(final Traversal.Admin<S , String> concatTraversal) {
+        final StringBuilder sb = new StringBuilder();
+        if (null != concatTraversal) {
+            while (concatTraversal.hasNext()) {
+                String result = concatTraversal.next();
```
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConcatStep.java:94]:
```suggestion
                final String result = concatTraversal.next();
```

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/BytecodeHelper.java [hunk 3]
```diff
@@ -196,6 +196,7 @@ public final class BytecodeHelper {
             put(GraphTraversal.Symbols.max, Arrays.asList(MaxGlobalStep.class, MaxLocalStep.class));
             put(GraphTraversal.Symbols.min, Arrays.asList(MinGlobalStep.class, MinGlobalStep.class));
             put(GraphTraversal.Symbols.mean, Arrays.asList(MeanGlobalStep.class, MeanLocalStep.class));
+            put(GraphTraversal.Symbols.concat,Collections.singletonList(ConstantStep.class));
```
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/BytecodeHelper.java:199]:
```suggestion
            put(GraphTraversal.Symbols.concat, Collections.singletonList(ConstantStep.class));
```

File: gremlin-dotnet/src/Gremlin.Net/Process/Traversal/GraphTraversal.cs [hunk 4]
```diff
@@ -501,6 +501,50 @@ private GraphTraversal(ICollection<ITraversalStrategy> traversalStrategies, Byte
...
+            var args = new List<object?>(concatStrings.Length);
+            args.AddRange(concatStrings);
+            Bytecode.AddStep("concat", args.ToArray());
+            return Wrap<TStart, String>(this);
+        }
+
+        /// <summary>
+        ///     Adds the concat step to this <see cref="GraphTraversal{SType, EType}" />.
+        /// </summary>
+        public GraphTraversal<TStart, String> Concat<String> (Scope scope, ITraversal concatTraversal)
```
Comment (vkagamlyk) [inline gremlin-dotnet/src/Gremlin.Net/Process/Traversal/GraphTraversal.cs:529]: is `concat` have `scope` now?

File: gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Concat.feature [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+@StepClassMap @StepConcat
+Feature: Step - concat()
+
+  #TODO - Gherkins unfortunately doesn't allow checking for empty strings, either pass null through or provide work around in test definition
+  @GraphComputerVerificationInjectionNotSupported
+  Scenario: g_injectXa_bX_concat
+    Given the modern graph
```
Comment (vkagamlyk) [inline gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Concat.feature:24]: better to use data from modern graph or use empty graph

File: gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Concat.feature [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+      | result |
+      | a |
+      | b |
+
+  @GraphComputerVerificationInjectionNotSupported
+  Scenario: g_injectXa_bX_concat_XcX
+    Given the modern graph
+    And the traversal of
+      """
+      g.inject("a", "b").concat("c");
```
[the remark this answers is not in the document]
Comment (vkagamlyk) [reply gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Concat.feature:40]: `g.inject("a", null).concat("c", null);`

File: gremlin-dotnet/src/Gremlin.Net/Process/Traversal/GraphTraversal.cs (hunk 4 above)
Comment (xiazcy) [reply gremlin-dotnet/src/Gremlin.Net/Process/Traversal/GraphTraversal.cs:529]: OH I thought I deleted that

File: gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Concat.feature (hunk 6 above)
[the remark this answers is not in the document]
Comment (xiazcy) [reply gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Concat.feature:40]: Ok, I will add them

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConcatStepTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        assertArrayEquals(new String[]{"ade", "bde", "cde"},
+                __.__("a", "b", "c").concat("d", "e").toList().toArray());
+        assertArrayEquals(new String[]{"abc", "bbc"},
+                __.__("a", "b").concat(__.__("b", "c")).toList().toArray());
+
+        assertArrayEquals(new String[]{"Mr.a", "Mr.b", "Mr.c", "Mr.d"},
+                __.__("a", "b", "c", "d").as("letters").constant("Mr.").concat(__.select("letters")).toList().toArray());
+
+        String nullStr = null;
+        assertNull(__.inject(null).concat(nullStr).next());
```
Comment (Cole-Greer) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ConcatStepTest.java:58]: Perhaps there should also be a test of 2 using empty strings

File: gremlin-language/src/main/antlr4/Gremlin.g4 [hunk 8]
```diff
@@ -802,6 +803,11 @@ traversalMethod_call
     | 'call' LPAREN stringArgument COMMA genericLiteralMapArgument COMMA nestedTraversal RPAREN #traversalMethod_call_string_map_traversal
     ;
 
+traversalMethod_concat
+	: 'concat' LPAREN RPAREN #traversalMethod_concat_Empty //is this needed given stringLiteralList?
```
Comment (Cole-Greer) [inline gremlin-language/src/main/antlr4/Gremlin.g4:807]: Appears to be redundant considering
```
| 'concat' LPAREN stringLiteralVarargs RPAREN #traversalMethod_concat_String
```
and
```
stringLiteralVarargs
    : (stringNullableArgument (COMMA stringNullableArgument)*)?
```

I would suggest removing this as long as we test parsing `concat()` via the grammar at some point.

**Comment by xiazcy** [comment]
Closing since it's opened to main repo
